### ISSUE TITLE
App\ResourceConnection\Proxy does not exist

### ISSUE THREAD

**Issue by khanzadimahdi** [magento/magento2#29197]
each time i run the the below command

```bash
bin/magento setup:di:compile
```
the response is `Class Magento\Framework\App\ResourceConnection\Proxy does not exist`

after that, all magento cli commands shows me the same error!!!

**Comment by bdt195** [comment]
Please make sure you have the correct permissions for the generated/code directory.
https://devdocs.magento.com/guides/v2.4/config-guide/prod/prod_file-sys-perms.html

**Comment by khanzadimahdi** [comment]
i  test it in my local system. all files have 777 permission.

```bash
sudo chmod -R 777 ./
```

i have changed all permissions to 777 , however this error persists.
magento2 terminal all commands generate this error and not work.

**Comment by engcom-Lima** [comment]
Hello @khanzadimahdi 
Could you provide more details for reproduce this issue?
Please add steps to reproduce and version of magento. Because I am not able to reproduce this issue on the latest 2.4-develop branch
![Screenshot_45](https://user-images.githubusercontent.com/60198022/88165670-ba020080-cc1e-11ea-909e-21d7193033ef.png)

Thank you!

**Comment by mcfriend99** [comment]
This issue still persists.

**Comment by engcom-Delta** [comment]
Hello @mcfriend99 
Could you provide more details for reproduce this issue?
Please add version of magento.

**Comment by travis-st** [comment]
I have the same issue on a fresh install when loading the /

I downloaded and installed `magento-ce-2.4.0_sample_data-2020-07-26-02-27-05.zip`. 

Ran 

```
php bin/magento setup:install --backend-frontname="admin" --session-save="files" --db-host="localhost" --db-name="magento2CE" --db-user="root" --db-password=****" --base-url="http://****/" --base-url-secure="https://***/" --admin-user="admin" --admin-password="***" --admin-email="***" --admin-firstname="***" --admin-lastname="***"

chmod a+w var/page_cache -R
chmod a+w var/cache -R
php bin/magento deploy:mode:set developer
tail /var/log/php-fpm/www-error.log -f -n 0
Stack trace:
#0 /var/www/html/magento2CE/vendor/magento/framework/Code/Reader/ClassReader.php(26): ReflectionClass->__construct('Magento\\Framewo...')
#1 /var/www/html/magento2CE/vendor/magento/framework/ObjectManager/Definition/Runtime.php(54): Magento\Framework\Code\Reader\ClassReader->getConstructor('Magento\\Framewo...')
#2 /var/www/html/magento2CE/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(48): Magento\Framework\ObjectManager\Definition\Runtime->getParameters('Magento\\Framewo...')
#3 /var/www/html/magento2CE/vendor/magento/framework/ObjectManager/ObjectManager.php(70): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('Magento\\Framewo...')
#4 /var/www/html/magento2CE/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(167): Magento\Framework\ObjectManager\ObjectManage in /var/www/html/magento2CE/vendor/magento/framework/Code/Reader/ClassReader.php on line 26

```

php-fpm v 7.319
nginx 1.18.0
Amazon Linux 2
mysqld  Ver 5.6.23 for Linux on x86_64 (MySQL Community Server (GPL))

Happy to provide any more info if you need it.

**Comment by travis-st** [comment]
Lol - why was this closed without comment? Seems to be a legit issue.

**Comment by khanzadimahdi** [comment]
I have closed this, because i cant regenerate it again.

**Comment by khanzadimahdi** [comment]
@travis-st okay i reopen it for you.

**Comment by travis-st** [comment]
I got it working, but then when I disabled TFA it happened again:

```
[root@ip-172-30-1-19 magento2CE]# php bin/magento module:disable Magento_TwoFactorAuth
The following modules have been disabled:
- Magento_TwoFactorAuth

Cache cleared successfully.
Generated classes cleared successfully. Please run the 'setup:di:compile' command to generate classes.
Info: Some modules might require static view files to be cleared. To do this, run 'module:disable' with the --clear-static-content option to clear them.
[root@ip-172-30-1-19 magento2CE]# tail /var/log/php-fpm/www-error.log -f -n 0
[30-Sep-2020 15:44:26 UTC] PHP Fatal error:  Uncaught ReflectionException: Class Magento\Framework\App\ResourceConnection\Proxy does not exist in /var/www/html/magento2CE/vendor/magento/framework/Code/Reader/ClassReader.php:26
Stack trace:
#0 /var/www/html/magento2CE/vendor/magento/framework/Code/Reader/ClassReader.php(26): ReflectionClass->__construct('Magento\\Framewo...')
#1 /var/www/html/magento2CE/vendor/magento/framework/ObjectManager/Definition/Runtime.php(54): Magento\Framework\Code\Reader\ClassReader->getConstructor('Magento\\Framewo...')
#2 /var/www/html/magento2CE/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(48): Magento\Framework\ObjectManager\Definition\Runtime->getParameters('Magento\\Framewo...')
#3 /var/www/html/magento2CE/vendor/magento/framework/ObjectManager/ObjectManager.php(70): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('Magento\\Framewo...')
#4 /var/www/html/magento2CE/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(167): Magento\Framework\ObjectManager\ObjectManage in /var/www/html/magento2CE/vendor/magento/framework/Code/Reader/ClassReader.php on line 26
^C
[root@ip-172-30-1-19 magento2CE]# php bin/magento setup:di:compile
Compilation was started.
App action list generation... 8/8 [============================] 100% 41 secs 372.0 MiB
Generated code and dependency injection configuration successfully.
[root@ip-172-30-1-19 magento2CE]# tail /var/log/php-fpm/www-error.log -f -n 0
[30-Sep-2020 15:46:55 UTC] PHP Fatal error:  Uncaught Magento\Framework\Exception\FileSystemException: Directory "/var/www/html/magento2CE/var/cache/" cannot be created Warning!mkdir(): Permission denied in /var/www/html/magento2CE/vendor/magento/framework/Filesystem/Driver/File.php:236
Stack trace:
#0 /var/www/html/magento2CE/vendor/magento/framework/Filesystem/Driver/File.php(210): Magento\Framework\Filesystem\Driver\File->mkdirRecursive('/var/www/html/m...', 511)
#1 /var/www/html/magento2CE/vendor/magento/framework/Filesystem/Directory/Write.php(94): Magento\Framework\Filesystem\Driver\File->createDirectory('/var/www/html/m...', 511)
#2 /var/www/html/magento2CE/vendor/magento/framework/App/Cache/Frontend/Factory.php(294): Magento\Framework\Filesystem\Directory\Write->create()
#3 /var/www/html/magento2CE/vendor/magento/framework/App/Cache/Frontend/Factory.php(132): Magento\Framework\App\Cache\Frontend\Factory->_getBackendOptions(Array)
#4 /var/www/html/magento2CE/vendor/magento/framework/App/Cache/Frontend/Pool.php(69): Magento\Framework\App\Cache\Fro in /var/www/html/magento2CE/vendor/magento/framework/Filesystem/Driver/File.php on line 236
^C
[root@ip-172-30-1-19 magento2CE]# mkdir -p var/cache
[root@ip-172-30-1-19 magento2CE]# chmod a+w var/cache -R
```

Basically this occurs if `setup:di:compile` isn't called?

**Comment by Dren7755** [comment]
I have the same problem right after composer install.
I have changed job company, workstation, project and this error is not absent. So, I can suppose that problem in magento.

**Comment by PascalBrouwers** [comment]
Able to reproduce by:
```
bin/magento c:f
composer install
bin/magento
```

Gives `Class Magento\Framework\App\ResourceConnection\Proxy does not exist`
Strangely enough running `composer install` again *after* the error seems to resolve it.
Somehow it needs the error.

**Comment by PascalBrouwers** [comment]
But, running a 
```
bin/magento c:f
composer install
```
Again after that and the error is back again

**Comment by adrian-martinez-interactiv4** [comment]
Thanks to the accurate explanation of @PascalBrouwers about how to reproduce it, I'd like to share here my findings.

**TL;DR**

This issue only seems to happen when:

- **PSR-0 autoload configuration includes `generated/code/`**

- **`generated/code/` folder is not empty and has generated code**, due to automatic generation, `setup:di:compile`, or `deploy:mode:set production`

**Error occurs when autoload classmap is generated using existing `generated/code/` files, and then files under `generated/code/` are removed** (See full explanation for further details).

 There are some ways to get rid of this behaviour. You can either (please note this options are mutually exclusive):

- **Run `bin/magento deploy:mode:set developer` prior to `composer install --optimize-autoloader`** to get `generated/code/` folder files cleared  (You may want to rerun `composer dump-autoload --optimize` after compilation to get classmap files updated)

- **Run `rm -rf generated/code/` to erase manually `generated/code/` folder files prior to `composer install --optimize-autoloader`** to get `generated/code/` folder files cleared  (You may want to rerun `composer dump-autoload --optimize` after compilation to get classmap files updated)

- **Run `rm -f var/.regenerate` to erase manually regenerate flag**, avoiding `generated/code/` being cleared at the start of any Magento request, including cli ones, **between `composer install --optimize-autoloader` and `bin/magento` command**:

```
composer install --optimize-autoloader
rm -f var/.regenerate
bin/magento list
```

- **Change the process strategy as follows:**

  - Update your composer.json, config section, define `"optimize-autoloader": false`, so autoloader optimization does not occur unless explicitly set in composer install / dump-autoload commands options
  - Run `composer install` without `--optimize-autoloader` flag
  - Do your work, possibly compiling files via `setup:di:compile` or similar
  - When your application has reached the desired state, run `composer dump-autoload --optimize` to generate classmap with all generated files

**Full Explanation**

Magento has a cleanup system that tries to detect when generated code and cache needs to be refreshed, **upon installation / uninstallation of Composer packages**. Whenever changes are detected, a file **flag is created: `var/.regenerate`** (please not this is not the same file as `var/.regenerate.lock`). At an very early stage of any Magento related code execution, the existence of this flag is checked, and **if it exists**, `\Magento\Framework\Code\GeneratedFiles::cleanGeneratedFiles` is called, so **`generated/code/` files are deleted** and expected to be regenerated dynamically.

The issue here is that **`magento/magento-composer-installer` Composer plugin fails to detect properly when there are actual changes in packages** (see `\MagentoHackathon\Composer\Magento\Plugin::onNewCodeEvent`), and **it sets `var/.regenerate` flag after each Composer install / update operation** (`post-install-cmd` /  `post-update-cmd` events), **regardless if any package has been installed or removed**. It also does not take care of `--optimize-autoloader` flags, or any flag at all.

Taking the provided replication example, process sequence is as follows (assuming existing files under `generated/code/` folder, see preconditions above):

- **`composer install --optimize-autoloader`**

  - Operation installs packages or not, it does not matter.
  - `vendor/composer/autoload_classmap.php` is generated, referencing existing files under `generated/code/` due to PSR-0 directive.
  - `magento/magento-composer-installer` sets request regeneration flag `var/.regenerate`

- **`bin/magento list`** (Fails)

  - Due to `var/.regenerate` existing, just after execution start, `generated/code/` files are deleted. Classmap under `vendor/composer/autoload_classmap.php` is not valid anymore, at least for those referencing `generated/code/` files.
  - Request execution resumes
  - Request fails to proceed properly, failing when it first references a just deleted class under generated/code/` (in this case `Magento\Framework\App\ResourceConnection\Proxy`, but it could be another one, problem is not really in that concrete class).

- **`composer install --optimize-autoloader`**

  - Operation installs packages or not, it does not matter.
  - `vendor/composer/autoload_classmap.php` is generated, with a difference from first time it was executed: as in previous `bin/magento list` command the `generated/code/` files were deleted due to regeneration request caused from 2 commands above, no references to `generated/code/` files are generated in classmap.
  - `magento/magento-composer-installer` sets request regeneration flag `var/.regenerate`


- **`bin/magento list`** (Succeeds)

  - Due to `var/.regenerate` existing, just after execution start, `generated/code/` files are deleted. Classmap under `vendor/composer/autoload_classmap.php` is still valid. Composer is forced to check filesystem and Magento code generator system generates non-existing classes, filling up again `generated/code/`.
  - Request execution resumes and succeeds.

At this point, you are at the start point, with files under `generated/code/`, and can indefinitely run this loop with same results (1 OK, 1 KO, 1 OK, 1 KO...).

**Comment by PascalBrouwers** [comment]
Yes, thank you for the full explanation. The problem is indeed switching between 2 folders who have a different classmap.
I went for option 2: "Run rm -rf generated/code/"
We have the issue that we are also using Redis so the classmap needs to be regenerated.

**Comment by adrian-martinez-interactiv4** [comment]
Yes, https://github.com/magento/magento2/blob/2.4-develop/lib/internal/Magento/Framework/Code/GeneratedFiles.php#L103 does not take care of cache configuration, and may not clean cache effectively, so in case you don't use built-in cache, you may need also to delete cache manually or via `cache:flush` before `rm -rf generated/code/`.

**Comment by obihill** [comment]
I also experienced this issue after a clean install of Magento 2.3.6 using Composer 1.10.13, and running `bin/magento list`.

The fix, in my case, was to reset permissions of the Magento root folder which was done using the following commands:

```
cd /var/www/html/magento2
sudo find var generated vendor pub/static pub/media app/etc -type f -exec chmod g+w {} +
sudo find var generated vendor pub/static pub/media app/etc -type d -exec chmod g+ws {} +
sudo chown -R :www-data .
sudo chmod u+x bin/magento
```

This code snippet is referenced in the [Magento Docs](https://devdocs.magento.com/guides/v2.3/install-gde/composer.html#set-file-permissions). You can also check these links for a better understanding of the commands: [How to use the chmod command](https://www.howtogeek.com/437958/how-to-use-the-chmod-command-on-linux/) and [Understanding the exec option of find](https://unix.stackexchange.com/questions/389705/understanding-the-exec-option-of-find).

**Comment by paul-uz** [comment]
I am hitting this error after doing the following steps:

* composer install
* set file/folder permissions
```
sudo find var generated vendor pub/static pub/media app/etc -type f -exec chmod u+w {} +
sudo find var generated vendor pub/static pub/media app/etc -type d -exec chmod u+w {} +
```
* run bin/magento setup:install ....

Based on the suggestions above, what would I need to do to rid me of this error?

**Comment by adrian-martinez-interactiv4** [comment]
I think `generated/code` is required to be as a PSR-0 entry, in order to allow Composer to know that it should also search in that directory for generated classes, otherwise it may won't be able to find generated files for autoloading them or including them in optimized autoload classmap when needed.

I've never run this error upon setup:install in a new, fresh created project. Without knowing your case in detail, I'd blindly suggest:

- Ensuring config.optimize-autoloader is not set or is explicitly set to false in composer.json.
- Ensuring config.classmap-authoritative is not set or is explicitly set to false in composer.json.
- Ensuring config.apcu-autoloader is not set or is explicitly set to false in composer.json.
- Ensuring your new installation does not share cache settings with another Magento installation.

If problem persists for you, would you be able to provide some more details?

**Comment by paul-uz** [comment]
@adrian-martinez-interactiv4 So I am attempting to do a "test" build using Github Actions.

I am pulling donw my code, installing composer dependencies, and then trying to run `setup:install` as such...

```
setup:install --admin-firstname="local" --admin-lastname="local" --admin-email="local@local.com" --admin-user="local" --admin-*** --base-url="http://magento.build/" --backend-frontname="admin" --db-host=localhost --db-name=magento --db-user=root --db-*** --use-secure=0 --use-rewrites=1 --use-secure-admin=0 --session-save="db" --currency="EUR" --language="en_US" --timezone="Europe/Rome" --cleanup-database --skip-db-validation
```

What is strange is a get mixed results. Without `sudo` I get the Proxy does not exist error. With `sudo` I eventually hit DB errors, such as `SQLSTATE[HY000] [2002] No such file or directory` or `SQLSTATE[HY000] [2054] The server requested authentication method unknown to the client` despite using correct credentials

**Comment by adrian-martinez-interactiv4** [comment]
Since Proxy does not exist error only happens without sudo, it looks like a permissions issue. I don't know how Github Actions manages users and permissions, but I'd check the following:
- Seems quite improbable for Github Actions, but ensure no actions are being performed using that repository code with another user than current, (i.e. web requests, cron execution, others), that may generate files with different users/permissions.
- Try with `generated/code` included as PSR-0 entry.
- Try without setting permissions explicitly
- Try changing owner instead of permissions (or both): `sudo chown -R "$(whoami):$(id -g -n $(whoami))" var generated vendor pub/static pub/media app/etc`

I don't know if I can be helpful further that this in your case.

**Comment by paul-uz** [comment]
@adrian-martinez-interactiv4 I am happy enough to use `sudo` but like i mentioned, I end up getting sporadic results when `setup:install` hits the DB creation step. For the time being, I have given up trying to run `setup:install` and instead I am just running `setup:di:compile` and `setup:static-content:deploy` to check it will build OK

**Comment by Asifjalil99** [comment]
@khanzadimahdi mostly this issues occur due to miss configuration of your folder users permission by permission i mean "**Owner**" and "**Group**"  e.g if your server using owner as "**root**" and group as "**root**" also and your files and folder user owner as "**user**" and group as "**server**" so you will face this error **Class Magento\Framework\App\ResourceConnection\Proxy does not exist**

**Conclusion**
your "**owner**" and "**group**" needs to be match with server.

**Comment by haiacom** [comment]
an other reason that cause the issue here is that there is no inode left for the current disk, you can just check by running this command "df -hi"
